File: Graphs Demo C#/Graphs Demo/Node.cs
namespace Graphs_Demo
{
    using System;
    using System.Collections.Generic;

    public class Node
    {
        private string name;
        private IList<Edge> edges;

        public Node(string name)
        {
            this.Name = name;
            this.edges = new List<Edge>();
        }

        public string Name
        {
            get
            {
                return this.name;
            }

            private set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Name cannot be an empty string!");
                }

                this.name = value;
            }
        }

        public IList<Edge> Edges
        {
            get
            {
                return this.edges;
            }
        }

        public void AddConnection(Node target, int weight, bool isTwoWay)
        {
            if (this == target)
            {
                throw new ArgumentException("Cannot add a connection to the same node!");
            }

            var edge = new Edge(target, weight);
            this.edges.Add(edge);

            if (isTwoWay)
            {
                var oppositeEdge = new Edge(this, weight);
                target.edges.Add(oppositeEdge);
            }
        }
    }
}

File: Graphs Demo C++/src/Node.cpp
#include "Node.h"
#include <Edge.h>
#include <algorithm>
#include <iostream>

Node::Node(string name)
{
    this->name = name;
}

Node::~Node()
{
}

Node::Node(const Node& other)
{
    this->name = other.name;
    this->edges = other.edges;
}

Node& Node::operator=(const Node& other)
{
    if (this != &other)
    {
        this->name = other.name;
        this->edges = other.edges;
    }

    return *this;
}

void Node::addConnection(Node& other, int weight, bool isTwoWay)
{
    Edge connection(other, weight);
    this->edges.push_back(connection);

    if (isTwoWay)
    {
        Edge reversedConnection(*this, weight);
        other.edges.push_back(reversedConnection);
    }
}

void printEdges(Edge& edge)
{
    cout << edge.getTarget().getName() << " ";
}

ostream& operator<<(ostream& os, Node& obj)
{
    os << obj.name;
    cout << " -> ";
    for_each(obj.edges.begin(), obj.edges.end(), printEdges);

    return os;
}

string& Node::getName()
{
    return this->name;
}

list<Edge> Node::getEdges() const
{
    return this->edges;
}
